### ISSUE TITLE
so why do we recreate a window each time we calculate loss

### ISSUE THREAD

**Issue by shi-yan** [OutofAi/2D-Gaussian-Splatting#13]
I feel the following code in the loss function redundant:

```python

 window = create_window(window_size, channel)
```

the window_size and channel are both fixed. create_window seems to generate a gaussian patch in a 11x11 patch

I feel that this can be cached .

I also have question regarding the SSIM loss. according to a paper I'm looking at https://arxiv.org/abs/2006.13846

the SIMM loss has 3 components, luminance, contrast and structure, whereas your implementation has only the first two, I wonder why? 

<img width="685" alt="Image" src="https://github.com/user-attachments/assets/c92e88b9-d32f-4deb-8da5-733878f37685" />

and the original equation has the alpha,beta exponents, but your version doesn't?

**Comment by OutofAi** [comment]
> the window_size and channel are both fixed. create_window seems to generate a gaussian patch in a 11x11 patch
> 
> I feel that this can be cached .

you could cache it but it's not a big performance overhead that would have a big effect on training I would have though

> the SIMM loss has 3 components, luminance, contrast and structure, whereas your implementation has only the first two, I wonder why?

and in terms of SSIM, in the paper you provided if you look at page 2 equation 6, that's the equation used which is 

![Image](https://github.com/user-attachments/assets/0e501f93-65aa-47f2-9dc9-bed6c27ba14e)

which comes from simplifying the individual terms for luminance, contrast, and structure into mathematical expressions that can be computed efficiently, which translates to this python code

```
SSIM_numerator = ((2*mu1_mu2 + C1)*(2*sigma12 + C2))
SSIM_denominator = ((mu1_sq + mu2_sq + C1)*(sigma1_sq + sigma2_sq + C2))
SSIM = SSIM_numerator / SSIM_denominator
```

**Comment by shi-yan** [comment]
Thank you very much! I'm trying to implement the same 2D gaussian splatting in cpp, I'm calculating derivatives by hand and using your code as a reference. Hopefully I can identify what's wrong.
